cheatsheets/Java_Security_Cheat_Sheet.md

```diff
@@ -0,0 +1,971 @@
+
+- **Never, ever write your own cryptographic functions.**
+- Wherever possible, try and avoid writing any cryptographic code at all. Instead try and either use pre-existing secret management solutions or the secret management solution provided by your cloud provider. For more information, see the [OWASP Secrets Management Cheat Sheet](/Secrets_Management_Cheat_Sheet).
+- Make sure your application or protocol can easily support a future change of cryptographic algorithms.
+- Use your package manager wherever possible to keep all of your packages up to date. Watch the updates on your development setup, and plan updates to your applications accordingly.
+- Wherever possible, use a trusted and well known implementation library rather than using the libraries built into JCA/JCE as it is far too easy to make cryptographic errors with them.
+- We will show examples below based on Google Tink which is an opinionated library for performing encryption operations.
+
+### Encryption for storage
+
+Follow the algorithm guidance in the [OWASP Cryptographic Storage Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Cryptographic_Storage_Cheat_Sheet.html#algorithms).
+
+#### Symmetric example using Google Tink
+
+Google Tink has documentation on performing common tasks.
+
+For example, this page shows how to perform simple symmetric encryption:
+
+[Simple Symmetric Encryption](https://developers.google.com/tink/encrypt-data) (From Google's website)
+
+The following code snippet shows an encapsulated use of this functionality:
+
+<details>
+  <summary>Click here to expand the code snippet</summary>
+
+``` java
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import com.google.crypto.tink.Aead;
+import com.google.crypto.tink.InsecureSecretKeyAccess;
+import com.google.crypto.tink.KeysetHandle;
+import com.google.crypto.tink.TinkJsonProtoKeysetFormat;
+import com.google.crypto.tink.aead.AeadConfig;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Base64;
+
+// AesGcmSimpleTest
+public class App {
+
+    // Based on example from:
+    // https://github.com/tink-crypto/tink-java/tree/main/examples/aead
+
+    public static void main(String[] args) throws Exception {
+        
+        // Key securely generated using:
+        // tinkey create-keyset --key-template AES128_GCM --out-format JSON --out aead_test_keyset.json
+          
+          
+          
+        // Register all AEAD key types with the Tink runtime.
+        AeadConfig.register();
+      
+        // Read the keyset into a KeysetHandle.
+        KeysetHandle handle =
+        TinkJsonProtoKeysetFormat.parseKeyset(
+            new String(Files.readAllBytes( Paths.get("/home/josh/aead_test_keyset.json")), UTF_8), InsecureSecretKeyAccess.get());
+
+        String message = "This message to be encrypted";
+        System.out.println(message);
+
+        String metadata = "Info about message";

```

Comment:
```java
        String metadata = "Info about message";
```

While this works on a very abstract level, I'm not sure developers will find this particularly useful. Even if this is straight from a Tink example, I don't think devs will find it particularly helpfil. Most often the AEAD metadata will be something that will be important to the recipient. E.g., 

```java
    String metadata = "Sender: tghosth";
```

or maybe
```java
    String metadata = "KeyID: 0xA4D4BB0785";  // indicate which key to use.
```

or maybe even something important about the message contents itself, such as:

```java
    String metadata = "Message expiration: 2023-06-21T13:34:56-05:00";
```

etc. It's something that might be useful to the recipient for processing, but it's confidentiality is not important. But ideally, try to give them ideas of what the AEAD metadata is typically used for instead of writing a general "Info about message".

Reply:
Great point, I have updated this to be more specific